### ISSUE TITLE
where in the world is the documentation for the REST api?

### ISSUE THREAD

**Issue by k1r0s** [saleslayer/Sales_Layer_PHP_SDK#15]
thx

https://api.saleslayer.com/

where i can find collections, auth management, filters, etc

**Comment by k1r0s** [comment]
in other words:

having a pair of CONECTOR_ID and SECRET_ID as X and Y...

what is the URL and params that I need in order to fetch a collection, say products. ONLY CURRENT DATA, not a history of changes, just the current data.

im looking for an asnwer like this:

curl -H "some_header: X Y" https://api.saleslayer.com/some_collection

**Comment by juliolaguna** [comment]
Hi Ciro.

Our Restful API defines two resources, one for retrieving data from Sales Layer platform and another one for inserting data into Sales Layer platform.

Our API resource for retrieving data, is based on the Connector definition and configuration, for setting response behavior and generation of results. Also the events over objects (aka. products, variants or categories) across the timeline are taken into account, and the behaviour is take from different parameters retrieved from GET/POST parameters in the request.

It's not a static API with defined fields for objects or collections where you can iterate over resultsets. There aren't exposed resources for theese kind of operations. 

You can check API Basic Working Concepts on our support center : https://support.saleslayer.com/api/introduction.

If you are looking for retrieving current data for your objects, you can always use **last_update** paramter set to 0 or simply don't send this parameter to the API request.

API Request parameter lists for retrieving data can be found here : https://support.saleslayer.com/api/other-modeling-parameters

Pay special attention to the contents of this URL explaining response behaviours for data extraction request on our API: https://support.saleslayer.com/api/api-operation-logic-data-extraction-mode-output

The API Resources I understand you're looking for like objects and collections must be programmed on your side based on retrieved data from our API.

In other hand the SDK provides a powerful tool for giving you the appropiate synchronization and storage aims. It's called SDK Updater and can be found as PHP Class in the SDK Package. This Open Source class provides the methods for API connection and storage on a MySQL Database. After define and configure your synchronization you can program your own Web Services based on your prefered technology , usually RESTful or SOAP. But also providing synchronized data to perform your own SQL Queries for your connected applications.

Below you can see the common Sales Layer Account integration model based on Web Services using SDK Updater.
![updater_model](https://user-images.githubusercontent.com/45973788/88026715-8d72b980-cb36-11ea-8767-b845ede82d44.png)


In short, basic query resources on particular objects or collections are outside the scope of our API. You can generate your own web services using the tools provided by the SDK or creating your own, based on the API responses.

Hope this helps you.

**Comment by k1r0s** [comment]
thanks for the answer. 

**however im missing the response schema.** 

My issue is that, when retrieving all the data I get a huge array with lots of properties and some items seems to be duplicated since I get the history of updates for every single item on the database. 

Even if last_update is not defined (or 0) I get every single version of that item through the history rather than the current one.

Im just looking to obtain a snapshot (current value) of every item in the database.

long story short; 

Given a PIM connector that exports the whole database. 

Can you please provide a CODE SAMPLE that fetches only the current state of every item and then iterates through the product data?

this is my code:

```php
<?php

require('./SalesLayer-Conn.php');

define("CONNECTOR_ID", "xxxxxxxxxxxx");
define("SECRET", "yyyyyyyyyyyyyy");

$SLconn = new SalesLayer_Conn(CONNECTOR_ID, SECRET);

// $date=date_create();
// $ts = date_timestamp_get($date);

$SLconn->get_info(0);

$arr = $SLconn->get_response_table_data();

function lookup($stack, $access) {
  foreach ($stack as $key => $value) {
    echo PHP_EOL;
    echo $access;
    if(is_array($value)) {
      lookup($value, $access . '||' . $key);
    }
  }
}

// lookup($arr['variants'], 'root');
// lookup($arr['categories'], 'root');
lookup($arr['products'], 'root');

```

My output is:

...
root||modified||2038||data
root||modified||2038||data
root||modified||2038||data
...

I need to know what is the response SCHEMA in order to properly iterate through AND im getting multiple versions for each product but i only need the latest

thanks for your patience!

**Comment by juliolaguna** [comment]
Hi Ciro.

Try with this piece of code (adapt it to your needs).

```php
$connector_id = 'XXXXXXXXX';
$secret_key = 'XXXXXXXXXX';

ini_set('memory_limit', '-1');
ini_set('max_execution_time', 0);

define('LOC_BASE', dirname(__FILE__) . '/');

include(LOC_BASE.'sdk-18-07-2020/src/SalesLayer-Conn.php'); //Modify with your own


$time = 1; 
//$time = file_get_contents('time.txt');

$msg_status = "";

//Create object
$sl_conn = new SalesLayer_Conn ($connector_id, $secret_key);

if ($sl_conn->has_response_error()) {
       throw new \InvalidArgumentException($sl_conn->get_response_error_message());
}

$sl_conn->set_group_multicategory(true);
//$sl_conn->set_first_level_parent_modifications(true);
//$sl_conn->set_same_parent_variants_modifications(true);
//$sl_conn->set_parents_category_tree(true);
$sl_conn->get_info(intval($time));
//$sl_conn->get_info();

$results = [];

$results['response_schema'] = $sl_conn->get_response_connector_schema();
$results['response_table_information'] = $sl_conn->get_response_table_information();
$results['response_table_data'] = $sl_conn->get_response_table_data();

$response_time = strtotime($sl_conn->get_response_time());

file_put_contents('time.txt', $response_time);

$msg_status .= "<p>Call time : " . date('d/m/Y H:i:s', $time) . " - timestamp : " . $time . "</p>";
$msg_status .= "<p>Response time : " . date('d/m/Y H:i:s', $response_time) . " - timestamp : " . $response_time . "</p>";

echo '<pre>' . print_r($results, true) . '</pre>';
```

Pay special attention to line:

`$sl_conn->set_group_multicategory(true);`

Probably it's the responsible of duplicated items, because by default API outputs a repeated record for each category a product belongs to.

This parameter groups a single output for each product and gives a category array under ID_catalogue key for product.

An important line too would be:

`$results['response_table_information'] = $sl_conn->get_response_table_information();`

On that response you can check for table schema and important information on fields.

-------------

I strongly recommend you to use a "Beautify Arrays" Utility (the one I use is https://github.com/uuf6429/nice_r/) and also test with few bunch of records by filtering output in the connector in the test phase.

In that way you can get testing results like this that will made your development process quick and easy

![array_test](https://user-images.githubusercontent.com/45973788/88053315-e1dc6000-cb5b-11ea-8e9b-ed15a0975f9c.png)

**Comment by juliolaguna** [comment]
![array_test_2](https://user-images.githubusercontent.com/45973788/88054684-0fc2a400-cb5e-11ea-9a10-ef93c03c79ab.png)

As you can see 

`$results['response_table_information'] = $sl_conn->get_response_table_information();`

will give you important information about table schema. This information will be useful for creating your own logic on iterations, or any kind of logic your custom app needs.

--------

Inspecting SDK code you'll be able to find lots of useful methods for getting almmost everything you need to interact with your Sales Layer Account Data.

Currently we're working on port all SDK and API documentation to our [Help Center](https://support.saleslayer.com/), but it's still in process. It will be available soon.

Hope it helps you.

**Comment by k1r0s** [comment]
Thanks @juliolaguna for the detailed and useful answer. Will report you back

**Comment by k1r0s** [comment]
the last question is how do I get the whole raw catalog. By default the API retrieves all data by modification date. Is there a way, for instance, to fetch all products in the database.

I mean, the **last_update** feature is nice but sometimes is not needed.

thanks!

**Comment by k1r0s** [comment]
Just to clarify. Is there a way to obtain all table rows ignoring which items were modified. Just retrieve the whole raw data

**Comment by juliolaguna** [comment]
Hi Ciro.

API always marks items as modified, also when new item is created. 

Just set the api time parameter as timestamp 1 or 0 as you just did:

**Call time : 01/01/1970 00:00:01 - timestamp : 1**

Via SDK Code:

`
$SLconn->get_info(0);
`

You see items under "M" subset, but because items where modified sometime after year 1970.

The amount of items API will output depends on connector configuration. Take a look at : https://support.saleslayer.com/connectors-general-aspects/connectors-general-aspects

You can output items depending on state, or some filters in connector configuration sections like output all or only visibles , etc ...

You can also get item creation date or item last modification date using connector formulas in the output data connector tab: GET_CREATION_DATE and GET_MODIFICATION_DATE formulas that may be useful for yor needs.

API behaviour is not complex but tricky, every time you make connector modifications (like adding a new field or modify a connector parameter like "Export items with status", etc ...), API outputs all filtered items (depending on connector filters and configuration) without taken into account time or last_time dates , the reason is for make sense on elements synchronization between new modified items and old items that your application would contain.

There are also features to take into account like API results cache, that sometimes could make you feel surprise when testing API results moving time dates in requests if no changes are made into the account , like items modificactions. But at now I think you haven't to fight with this behaviour.

Try the code I give you posts ago, with time 1, and check the resultsets. Don't worry about items under "modified" section, every item to take into account is under this section. It's the item current data !!

**Comment by juliolaguna** [comment]
Take a look at:

https://support.saleslayer.com/api/parameterization-export

For knowing how to parametrize the connector output via API response.

and also take a look at:

https://support.saleslayer.com/api/api-operation-logic-data-extraction-mode-output

where you can get a basic idea of API timeline behaviour.

**Comment by k1r0s** [comment]
Thanks! will report you back
